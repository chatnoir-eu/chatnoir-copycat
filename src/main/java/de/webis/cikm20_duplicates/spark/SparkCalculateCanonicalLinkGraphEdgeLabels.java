package de.webis.cikm20_duplicates.spark;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import org.apache.htrace.shaded.fasterxml.jackson.databind.ObjectMapper;
import org.apache.spark.HashPartitioner;
import org.apache.spark.Partitioner;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.google.common.collect.ImmutableList;

import de.webis.cikm20_duplicates.spark.SparkCanonicalLinkGraphExtraction.CanonicalLinkGraphEdge;
import de.webis.cikm20_duplicates.spark.eval.SparkAnalyzeCanonicalLinkGraph;
import lombok.SneakyThrows;
import scala.Tuple2;

public class SparkCalculateCanonicalLinkGraphEdgeLabels {

	private static final String DIR = "cikm2020/canonical-link-graph/";
	
	public static void main(String[] args) {
		String[] corpora = new String[] {/*"cw09" ,*/ "cw12"/*, "cc-2015-11"*/};
		
		try (JavaSparkContext context = context()) {
			for(String corpus : corpora) {
				JavaRDD<String> input = context.textFile(DIR + corpus);
				
				edgeLabels(input, new HashPartitioner(10000))
					.saveAsTextFile(DIR + corpus + "-calulated-edges");
			}
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/analyze-canonical-link-graph");
	
		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> edgeLabels(JavaRDD<String> input, Partitioner partitioner) {
		return input.mapToPair(i -> toPair(i))
			.filter(i -> i != null)
			.repartitionAndSortWithinPartitions(partitioner)
			.groupByKey()
			.flatMapToPair(i -> pairsForCalculation(i))
			.filter(i -> i != null && i._1() != null && i._2() != null && i._2()._1() != null && i._2()._2() != null)
			.repartitionAndSortWithinPartitions(partitioner)
			.map(i -> reportEdgeOrNull(i._2()._1(), i._2()._2()))
			.filter(i -> i != null);
	}

	private static Tuple2<String, CanonicalLinkGraphEdge> toPair(String src) {
		CanonicalLinkGraphEdge edge = CanonicalLinkGraphEdge.fromString(src);
		String url = edge.getCanonicalLink().toString();
		String host = SparkAnalyzeCanonicalLinkGraph.hostFromUrl(url);
		if(host == null || host.equalsIgnoreCase("ERROR-PARSING-URL")) {
			return null;
		}
		
		return new Tuple2<>(url, edge);
	}
	
	private static Iterator<Tuple2<String, Tuple2<CanonicalLinkGraphEdge, CanonicalLinkGraphEdge>>> pairsForCalculation(Tuple2<String, Iterable<CanonicalLinkGraphEdge>> bla) {
		List<CanonicalLinkGraphEdge> edges = new ArrayList<>(ImmutableList.copyOf(bla._2().iterator()));
		Collections.sort(edges, (a,b) -> a.getDoc().getId().compareTo(b.getDoc().getId()));
		
		Stream<Tuple2<Integer, Integer>> indizesToCompare = IntStream.range(0, edges.size()).mapToObj(i -> i)
				.flatMap(i -> IntStream.range(i+1, edges.size()).mapToObj(j -> new Tuple2<Integer, Integer>(i,j)));
	
		return indizesToCompare.map(i -> new Tuple2<>(
				edges.get(i._1()).getDoc().getId() +"-" +edges.get(i._2()).getDoc().getId(),
				new Tuple2<>(edges.get(i._1()), edges.get(i._2()))
		)).iterator();
	}

	@SneakyThrows
	private static String reportEdgeOrNull(CanonicalLinkGraphEdge a, CanonicalLinkGraphEdge b) {
		if(!a.getCanonicalLink().equals(b.getCanonicalLink()) || a.getDoc().getId().compareTo(b.getDoc().getId()) >= 0) {
			return null;
		}
		
		Map<String, Object> ret = new LinkedHashMap<>();
		ret.put("canonicalLink", a.getCanonicalLink());
		ret.put("firstDoc", a);
		ret.put("secondDoc", b);
		ret.put("s3score", SparkEnrichRelevanceTransferPairs.s3Score(a.getDoc(), b.getDoc()));
		
		return new ObjectMapper().writeValueAsString(ret);
	}
}
