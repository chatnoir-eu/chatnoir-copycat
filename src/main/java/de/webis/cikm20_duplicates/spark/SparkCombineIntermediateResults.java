package de.webis.cikm20_duplicates.spark;

import java.util.Map;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.SneakyThrows;

public class SparkCombineIntermediateResults {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> input = context.textFile("cikm2020/near-duplicate-tasks-cw09-cw12");
			
			combineIntermediateResults(input, 5000)
				.saveAsTextFile("cikm2020/results/test-01");
		}
	}

	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/candidates");

		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> combineIntermediateResults(JavaRDD<String> inputRDD, int partitions) {
		return inputRDD
				.map(i -> toCsv(i))
				.distinct()
				.sortBy(i -> i, Boolean.TRUE, partitions);
	}

	@SneakyThrows
	@SuppressWarnings("unchecked")
	private static String toCsv(String json) {
		Map<String, Object> ret = new ObjectMapper().readValue(json, Map.class);
		String firstId = (String) ret.get("firstId");
		String secondId = (String) ret.get("secondId");
		Integer hemmingDistance = (Integer) ret.get("hemmingDistance");
		
		if(firstId == null|| secondId == null || hemmingDistance == null) {
			throw new IllegalArgumentException("");
		}
		
		return firstId + "," + secondId + "," + hemmingDistance;
	}
}
