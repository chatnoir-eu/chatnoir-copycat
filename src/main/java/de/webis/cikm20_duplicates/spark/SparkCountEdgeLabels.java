package de.webis.cikm20_duplicates.spark;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.math.BigIntegerMath;

import lombok.SneakyThrows;
import scala.Tuple2;

public class SparkCountEdgeLabels {

	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> exactDuplicates = context.textFile("cikm2020/exact-duplicates-simhash-cw09-cw12");
			JavaRDD<String> pairs = context.textFile("cikm2020/results/test-01");
		
			countEdgeLabels(exactDuplicates, pairs).saveAsTextFile("cikm2020/near-duplicate-graph/edge-count-cw09-cw12");
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/count-edges");

		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> countEdgeLabels(JavaRDD<String> exactDuplicates, JavaRDD<String> pairs) {
		JavaPairRDD<String, BigInteger> a = exactDuplicates.flatMapToPair(i -> flatten(i).iterator())
				.reduceByKey((i,j) -> i.add(j), 10)
				.filter(i -> !i._2.equals(BigInteger.ZERO));
		
		JavaPairRDD<String, BigInteger> b = pairs.mapToPair(i -> toTuple(i))
				.reduceByKey((i,j) -> i.add(j), 10)
				.filter(i -> !i._2.equals(BigInteger.ZERO));
		
		return a.union(b)
				.reduceByKey((i,j) -> i.add(j), 10)
				.filter(i -> !i._2.equals(BigInteger.ZERO))
				.map(i -> "{\"" + i._1() + "\":" + i._2() + "}");
	}

	private static Tuple2<String, BigInteger> toTuple(String a) {
		String[] parsed = a.split(",");
		if(parsed.length != 3) {
			throw new RuntimeException("Can not transform input to tuple: '" + a + "'");
		}
		
		String name;
		
		if((parsed[0].startsWith("clueweb09") && parsed[1].startsWith("clueweb12")) || (parsed[0].startsWith("clueweb12") && parsed[1].startsWith("clueweb09"))) {
			name = "clueweb09<->clueweb12";
		} else if (parsed[0].startsWith("clueweb09") && parsed[1].startsWith("clueweb09")) {
			name = "clueweb09<->clueweb09";
		} else if (parsed[0].startsWith("clueweb12") && parsed[1].startsWith("clueweb12")) {
			name = "clueweb12<->clueweb12";
		} else {
			throw new RuntimeException("Can not handle '" + a + "'.");
		}
		
		return new Tuple2<>(name +" (k=" + Integer.parseInt(parsed[2]) + ")", BigInteger.ONE);
	}
	
	@SneakyThrows
	@SuppressWarnings({ "unchecked", "rawtypes" })
	private static List<Tuple2<String, BigInteger>> flatten(String src) {
		Map<String, Object> s = new ObjectMapper().readValue(src, Map.class);
		List<String> docs = (List) s.get("equivalentDocuments");
		
		BigInteger all = edgesBetweenFullyConnectedGraph(docs.size());
		BigInteger clueWeb09 = edgesBetweenFullyConnectedGraph(docs.stream().filter(i -> i.startsWith("clueweb09")).count());
		BigInteger clueWeb12 = edgesBetweenFullyConnectedGraph(docs.stream().filter(i -> i.startsWith("clueweb12")).count());
		
		return Arrays.asList(
			new Tuple2<>("clueweb09<->clueweb09 (k=0)", clueWeb09),
			new Tuple2<>("clueweb12<->clueweb12 (k=0)", clueWeb12),
			new Tuple2<>("clueweb09<->clueweb12 (k=0)", all.subtract(clueWeb12).subtract(clueWeb09))
		);
	}
	
	private static BigInteger edgesBetweenFullyConnectedGraph(long count) {
		if(count <= 1) {
			return BigInteger.ZERO;
		}
		
		BigInteger nodeCount = BigIntegerMath.factorial((int) count);
		BigInteger denominator = BigInteger.valueOf(2l).multiply(BigIntegerMath.factorial(((int)count) -2));
		
		return nodeCount.divide(denominator);
	}
	
	
}
