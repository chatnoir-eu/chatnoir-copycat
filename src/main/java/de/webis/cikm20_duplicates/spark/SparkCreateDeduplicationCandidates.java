package de.webis.cikm20_duplicates.spark;

import java.io.Serializable;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.curator.shaded.com.google.common.collect.Iterators;
import org.apache.htrace.shaded.fasterxml.jackson.databind.ObjectMapper;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.google.common.hash.BloomFilter;
import com.google.common.hash.Funnels;

import de.webis.cikm20_duplicates.util.FingerPrintUtil.Fingerprinter;
import de.webis.cikm20_duplicates.util.SourceDocuments.CollectionDocumentWithTopics;
import de.webis.cikm20_duplicates.util.SourceDocuments.DocumentWithFingerprint;
import de.webis.trec_ndd.trec_collections.AnseriniCollectionReader;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.collection.ClueWeb09Collection.Document;
import lombok.SneakyThrows;
import scala.Tuple2;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
public class SparkCreateDeduplicationCandidates {

	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints");
			
//			JavaRDD<String> duplicationCandidates = duplicationCandidatesFromFingerprints(input)
//				.map(i -> i.toString());
//			
//			duplicationCandidatePairsFromFingerprints(duplicationCandidates)
//				.saveAsTextFile("cikm2020/candidate-pairs");
			
			hashPartitionToDocument(input, DeduplicationStrategy.MIN_HASH_DEDUPLICATION_STRATEGY)
				.aggregateByKey(0l, (count, doc) -> Long.valueOf((long)(count +1)), (i,j) -> Long.valueOf((long) i+j))
				.map(i -> "{\"hash-partition\": "+ i._1() +", \"count\": "+ i._2() +"}")
				.saveAsTextFile("cikm2020/count-per-hash-partition");
		}
	}
	
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/candidates");

		return new JavaSparkContext(conf);
	}
	
	public static JavaPairRDD<String, Iterable<Tuple2<String, CollectionDocument>>> topicsToImportantDocuments(JavaRDD<String> rdd) {
		JavaRDD<Tuple2<String, CollectionDocument>> ret = rdd
				.map(i -> CollectionDocumentWithTopics.fromString(i))
				.flatMap(i -> flattenTopicsForDoc(i));
 
		return ret.groupBy(i -> i._1());
	}

	private static Iterator<Tuple2<String, CollectionDocument>> flattenTopicsForDoc(CollectionDocumentWithTopics docWithTopics) {
		return docWithTopics.getTopics().stream()
				.map(topic -> new Tuple2<>(topic, docWithTopics.getDoc()))
				.iterator();
	}

	public static <T extends Comparable<T>> JavaRDD<Tuple2<String, Set<T>>> topicsToFingerPrintsOfImportantDocsPerTopic(JavaRDD<String> rdd, Fingerprinter<T> fingerprinter) {
		return topicsToImportantDocuments(rdd)
				.map(i -> new Tuple2<>(i._1, combineAll(i._2, fingerprinter)));
	}

	private static <T extends Comparable<T>> Set<T> combineAll(Iterable<Tuple2<String, CollectionDocument>> a, Fingerprinter<T> fingerprinter) {
		Set<T> ret = new HashSet<>();
		Iterator<CollectionDocument> docIterator = Iterators.transform(a.iterator(), i -> i._2());
		docIterator.forEachRemaining(doc -> ret.addAll(fingerprinter.fingerprint(doc)));

		return ret;
	}

	public static JavaRDD<Tuple2<String, CollectionDocument>> candidatesForAllSourceDocuments(JavaSparkContext context, JavaRDD<String> sourceDocuments, Fingerprinter<Integer> fingerprinter, AnseriniCollectionReader<Document> acr) {
		Map<String, BloomFilter<Integer>> topicBloomFilters = new HashMap<>();
		
		Map<String, Set<Integer>> topicToFingerPrintUnits = topicsToFingerPrintsOfImportantDocsPerTopic(sourceDocuments, fingerprinter)
				.collect().stream()
				.collect(Collectors.toMap(i -> i._1(), i -> i._2()));
		int overallElements = topicToFingerPrintUnits.values().stream().mapToInt(i -> i.size()).sum();
		
		BloomFilter<Integer> bf = BloomFilter.create(Funnels.integerFunnel(), overallElements + 100000, 1.0e-8);

		for(Entry<String, Set<Integer>> topicToFingerpintUnit : topicToFingerPrintUnits.entrySet()) {
			BloomFilter<Integer> topicBf = BloomFilter.create(Funnels.integerFunnel(), topicToFingerpintUnit.getValue().size() + 10000, 1.0e-8);
			
			for(Integer fingerprintUnit: topicToFingerpintUnit.getValue()) {
				topicBf.put(fingerprintUnit);
				bf.put(fingerprintUnit);
			}
			
			topicBloomFilters.put(topicToFingerpintUnit.getKey(), topicBf);
		}
		
		JavaRDD<CollectionDocument> allDocs = context.parallelize(acr.segmentPaths())
			.flatMap(s -> acr.collectionDocumentsInPath(s));
		
		return allDocs.flatMap(doc -> docToCandidateTopics(fingerprinter, doc, bf, topicBloomFilters));
	}
	
	private static <T extends Comparable<T>> Iterator<Tuple2<String, CollectionDocument>> docToCandidateTopics(Fingerprinter<T> fingerprinter, CollectionDocument doc, BloomFilter<T> bf, Map<String, BloomFilter<T>> topicBloomFilters) {
		List<T> fingerPrint = fingerprinter.fingerprint(doc);
		Set<String> candidateTopics = new HashSet<>();
		
		for(T fingerPrintUnit: fingerPrint) {
			if(bf.mightContain(fingerPrintUnit)) {
				for(Entry<String, BloomFilter<T>> e : topicBloomFilters.entrySet()) {
					if(e.getValue().mightContain(fingerPrintUnit)) {
						candidateTopics.add(e.getKey());
					}
				}
			}
		}
		
		return candidateTopics.stream()
				.map(i -> new Tuple2<>(i, doc))
				.iterator();
	}

	private static JavaPairRDD<Integer, String> hashPartitionToDocument(JavaRDD<String> docsWithFingerprint, DeduplicationStrategy f) {
		return docsWithFingerprint
				.map(i -> DocumentWithFingerprint.fromString(i))
				.flatMapToPair(doc -> extractHashesToDocId(doc, f));
	}
	
	public static JavaRDD<String> duplicationCandidatePairsFromFingerprints(JavaRDD<String> docsWithFingerprint, DeduplicationStrategy f) {
		JavaPairRDD<Integer, String> parsedInput = hashPartitionToDocument(docsWithFingerprint, f);
		
		return parsedInput.join(parsedInput)
				.map(i -> emitPairOrNull(i._2()))
				.filter(i -> i != null)
				.distinct();
	}
	
	public static List<Integer> useMinHash(DocumentWithFingerprint doc) {
		return doc.getMinHashParts();
	}
	
	public static List<Integer> useSimHash(DocumentWithFingerprint doc) {
		return doc.getSimHash65BitParts();
	}
	
	@SneakyThrows
	private static String emitPairOrNull(Tuple2<String, String> pair) {
		if(pair == null || pair._1() == null || pair._2() == null || pair._1().compareTo(pair._2()) >= 0) {
			return null;
		}

		Map<String, String> candidate = new LinkedHashMap<>();
		candidate.put("firstId", pair._1());
		candidate.put("secondId", pair._2());
				
		return new ObjectMapper().writeValueAsString(candidate);
	}

	private static Iterator<Tuple2<Integer, String>> extractHashesToDocId(DocumentWithFingerprint doc, DeduplicationStrategy f) {
		return f.extract(doc).stream()
				.map(hash -> new Tuple2<>(hash, doc.getDocId()))
				.iterator();
	}
	
	public static JavaRDD<DocumentWithFingerprint> duplicationCandidatesFromFingerprints(JavaRDD<String> docsWithFingerprint) {
		JavaRDD<DocumentWithFingerprint> parsedInput = docsWithFingerprint.map(i -> DocumentWithFingerprint.fromString(i));
		BloomFilter<Integer> bf = bf(parsedInput);
		
		return parsedInput.filter(doc -> doc.getMinHashParts().stream().anyMatch(i -> bf.mightContain(i)));
	}
	
	private static BloomFilter<Integer> bf(JavaRDD<DocumentWithFingerprint> docs) {
		List<Integer> allElements = docs.filter(i -> SparkCreateSourceDocuments.DOCS_TO_TOPIC.containsKey(i.getDocId()))
			.flatMap(i -> i.getMinHashParts().iterator())
			.distinct()
			.collect();
		
		BloomFilter<Integer> ret = BloomFilter.create(Funnels.integerFunnel(), allElements.size(), 1.0e-8);
		allElements.forEach(i -> ret.put(i));
		
		return ret;
	}
	
	static interface DeduplicationStrategy extends Serializable {
		public List<Integer> extract(DocumentWithFingerprint doc);
		
		@SuppressWarnings("serial")
		public static final DeduplicationStrategy MIN_HASH_DEDUPLICATION_STRATEGY = new DeduplicationStrategy() {
			@Override
			public List<Integer> extract(DocumentWithFingerprint doc) {
				return doc.getMinHashParts();
			}
		};
		
		@SuppressWarnings("serial")
		public static final DeduplicationStrategy SIM_HASH_DEDUPLICATION_STRATEGY = new DeduplicationStrategy() {
			@Override
			public List<Integer> extract(DocumentWithFingerprint doc) {
				return doc.getSimHash65BitParts();
			}
		};
	}
}
