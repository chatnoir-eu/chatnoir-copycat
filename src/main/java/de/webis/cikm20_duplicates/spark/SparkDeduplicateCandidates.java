package de.webis.cikm20_duplicates.spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import de.webis.cikm20_duplicates.util.ClientLocalDeduplication;
import de.webis.cikm20_duplicates.util.ClientLocalDeduplication.DeduplicationTask;
import scala.Tuple2;

/**
 * 
 * @author Maik FrÃ¶be
 *
 */
public class SparkDeduplicateCandidates {
//	public static void main(String[] args) {
//		try (JavaSparkContext context = context()) {
//			context.textFile("cikm2020/near-duplicate-tasks-cw09-cw12")
//				.map(i-> DeduplicationTask.fromString(i))
//				.flatMap(i -> ClientLocalDeduplication.fullDeduplication(i.getEntries()).iterator())
//				.saveAsTextFile("cikm2020/deduplication/near-duplicates/cw09-cw12");
//		}
//	}
	
	public static void main(String[] args) {
		String corpus = "cw09-cw12";
		try (JavaSparkContext context = context()) {
			JavaPairRDD<Integer, Integer> deduplicationTaskSizeToCount = context.textFile(SparkCreateDeduplicationCandidates.inputPath(corpus))
				.mapToPair(i-> new Tuple2<>(DeduplicationTask.fromString(i).getEntries().size(), 1))
				.reduceByKey((a,b) -> a+b);
			
			deduplicationTaskSizeToCount
				.map(i -> "{\"groupSize\":" + i._1() + ",\"count\":" + i._2() + "}")
				.saveAsTextFile("cikm2020/document-fingerprints-final/deduplication-task-size-to-count/" + corpus);
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/deduplication/near-duplicates");

		return new JavaSparkContext(conf);
	}
}
