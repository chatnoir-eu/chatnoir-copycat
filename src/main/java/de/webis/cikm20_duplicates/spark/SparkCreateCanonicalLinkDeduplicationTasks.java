package de.webis.cikm20_duplicates.spark;

import java.io.Serializable;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.apache.spark.Partitioner;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.collect.ImmutableList;

import de.webis.cikm20_duplicates.spark.SparkCreateDeduplicationCandidates.DeduplicationUnit;
import de.webis.cikm20_duplicates.spark.SparkCreateDeduplicationCandidates.IntPartitioner;
import de.webis.cikm20_duplicates.util.ClientLocalDeduplication.DeduplicationTask;
import de.webis.cikm20_duplicates.util.SourceDocuments.DocumentWithFingerprint;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.SneakyThrows;
import scala.Tuple2;

public class SparkCreateCanonicalLinkDeduplicationTasks {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			for(String corpus: new String[] {"cw12"/*, "cc-2015-11", "cc-2017-04"*/}) {
				JavaRDD<String> input = context.textFile(inputPath(corpus));
				
				urlDeduplicationTask(input, new IntPartitioner(50000))
					.saveAsTextFile(path(corpus) + "-near-duplicate-tasks");
			}
		}
	}
	
	public static String inputPath(String corpus) {
		if ("cw09-cw12".equals(corpus)) {
			return "cikm2020/document-fingerprints-final/cw*-jsonl.bzip2";
		} else if ("cw09-cw12-cc-2015-11".equals(corpus)) {
			return "cikm2020/document-fingerprints-final/{cw,cc-2015}*-jsonl.bzip2";
		}
		
		return "cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2";
	}
	
	public static String path(String corpus) {
		return "cikm2020/deduplication-final/64BitK3SimHashOneGramms-canonical-urls/" + corpus;
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/canonical-link-deduplication-tasks");

		return new JavaSparkContext(conf);
	}

	public static JavaRDD<String> urlDeduplicationTask(JavaRDD<String> input, Partitioner partitioner) {
		return hashPartitionToDocument(input).groupByKey(partitioner)
				.flatMap(i -> workingPackages2(i._2()));
	}

	private static Iterator<String> workingPackages2(Iterable<DeduplicationUnit2> bla) {
		List<DeduplicationUnit2> tasks = new LinkedList<>(new HashSet<>(ImmutableList.copyOf(bla)));
		Map<URL, List<DeduplicationUnit>> ret = new LinkedHashMap<>();
		
		for(DeduplicationUnit2 t: tasks) {
			if(!ret.containsKey(t.url)) {
				ret.put(t.url, new ArrayList<>());
			}
			
			ret.get(t.url).add(new DeduplicationUnit(t.id, t.hashParts));
		}
		
		return ret.values().stream()
				.filter(i -> i.size() > 1)
				.map(i -> new DeduplicationTask(i).toString())
				.iterator();
	}

	private static JavaPairRDD<String, DeduplicationUnit2> hashPartitionToDocument(JavaRDD<String> docsWithCanonicalURL) {
		return docsWithCanonicalURL
				.flatMapToPair(doc -> extractCanonicalUrlToDocId(doc));
	}

	private static Iterator<Tuple2<String, DeduplicationUnit2>> extractCanonicalUrlToDocId(String src) {
		DocumentWithFingerprint doc = DocumentWithFingerprint.fromString(src);
		
		if(doc == null || doc.getCanonicalURL() == null ||doc.getCanonicalURL().toString().trim().isEmpty()) {
			return Collections.emptyIterator();
		}
		
		DeduplicationUnit2 dedupUnit = new DeduplicationUnit2(doc.getDocId(), doc.getCanonicalURL(), doc.getFingerprints().get("64BitK3SimHashOneGramms"));
		return Arrays.asList(new Tuple2<>(doc.getCanonicalURL().toString(), dedupUnit)).iterator();
	}
	
	private static Iterator<Tuple2<Integer, DeduplicationUnit2>> extractHashesToDocId(String src) {
		DocumentWithFingerprint doc = DocumentWithFingerprint.fromString(src);
		List<Tuple2<Integer, DeduplicationUnit2>> ret = new ArrayList<>();
		
		if(doc.getCanonicalURL() == null ||doc.getCanonicalURL().toString().trim().isEmpty()) {
			return ret.iterator();
		}
		
		DeduplicationUnit2 dedupUnit = new DeduplicationUnit2(doc.getDocId(), doc.getCanonicalURL(), doc.getFingerprints().get("64BitK3SimHashOneGramms"));
		
		for(Integer hashPart: dedupUnit.getHashParts()) {
			ret.add(new Tuple2<>(hashPart, dedupUnit));
		}
		
		return ret.iterator();
	}
	
	@Data
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class DeduplicationUnit2 implements Serializable {
		private String id;
		private URL url;
		private ArrayList<Integer> hashParts;
		
		@Override
		@SneakyThrows
		public String toString() {
			return new ObjectMapper().writeValueAsString(this);
		}
	}
}
