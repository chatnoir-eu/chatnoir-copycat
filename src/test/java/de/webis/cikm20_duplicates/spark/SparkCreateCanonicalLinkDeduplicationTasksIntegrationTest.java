package de.webis.cikm20_duplicates.spark;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import org.apache.spark.HashPartitioner;
import org.apache.spark.api.java.JavaRDD;
import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;

import com.holdenkarau.spark.testing.SharedJavaSparkContext;

public class SparkCreateCanonicalLinkDeduplicationTasksIntegrationTest extends SharedJavaSparkContext {
	
	@Test
	public void testDeduplicationTasksOfIdenticalDocumentsForDifferentUrls() {
		List<String> expected = Collections.emptyList();
		List<String> actual = deduplicationTasks(
			doc("1", "1", Arrays.asList(1,2,3,4)),
			doc("2", "2", Arrays.asList(1,2,3,4)),
			doc("3", "3", Arrays.asList(1,2,3,4)),
			doc("4", "4", Arrays.asList(1,2,3,4))
		);
		
		Assert.assertEquals(expected, actual);
	}
	
	@Test
	public void testDeduplicationTasksOfDifferentDocumentsForIdenticalUrls() {
		List<String> expected = Collections.emptyList();
		List<String> actual = deduplicationTasks(
			doc("1", "1", Arrays.asList(1,2,3,4)),
			doc("1", "1", Arrays.asList(5,6,7,8)),
			doc("1", "1", Arrays.asList(9,10,11,12)),
			doc("1", "1", Arrays.asList(13,14,15,16))
		);
		
		Assert.assertEquals(expected, actual);
	}
	
	@Test
	public void testDeduplicationTasksOfSomeIdenticalDocumentsForSomeIdenticalUrls() {
		List<String> actual = deduplicationTasks(
			doc("1", "1", Arrays.asList(1,2,3,4)),
			doc("2", "2", Arrays.asList(1,2,3,4)),
			doc("3", "3", Arrays.asList(1,2,3,4)),
			doc("4", "4", Arrays.asList(1,2,3,4)),
			doc("5", "1", Arrays.asList(1,2,3,5)),
			doc("6", "1", Arrays.asList(1,2,3,6))
		);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void testDeduplicationTasksOfSomeIdenticalDocumentsForSomeIdenticalUrlsAndSomeDuplicates() {
		List<String> actual = deduplicationTasks(
			doc("1", "1", Arrays.asList(1,2,3,4)),
			doc("2", "2", Arrays.asList(1,2,3,4)),
			doc("6", "1", Arrays.asList(1,2,3,6)),
			doc("6", "1", Arrays.asList(1,2,3,6)),
			doc("3", "3", Arrays.asList(1,2,3,4)),
			doc("4", "4", Arrays.asList(1,2,3,4)),
			doc("5", "1", Arrays.asList(1,2,3,5)),
			doc("6", "1", Arrays.asList(1,2,3,6)),
			doc("2", "2", Arrays.asList(1,2,3,4)),
			doc("2", "2", Arrays.asList(1,2,3,4)),
			doc("4", "4", Arrays.asList(1,2,3,4))
		);
		
		Approvals.verifyAsJson(actual);
	}
	
	private List<String> deduplicationTasks(String...docs) {
		JavaRDD<String> input = jsc().parallelize(Arrays.asList(docs));
		JavaRDD<String> output = SparkCreateCanonicalLinkDeduplicationTasks.urlDeduplicationTask(input, new HashPartitioner(1));
		List<String> ret = new ArrayList<>(output.collect());
		Collections.sort(ret);
		
		return ret;
	}
	
	private static String doc(String id, String canonicalUrl, List<Integer> hashes) {
		return "{\"docId\": \"" + id + "\",\"canonicalURL\":\"https://" + canonicalUrl + ".de\",\"fingerprints\":{\"64BitK3SimHashOneGramms\": " + hashes +"}}";
	}
}
