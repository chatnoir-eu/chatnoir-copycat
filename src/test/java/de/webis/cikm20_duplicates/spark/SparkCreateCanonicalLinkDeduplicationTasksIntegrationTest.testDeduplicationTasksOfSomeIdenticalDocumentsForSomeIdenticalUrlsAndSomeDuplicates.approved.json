[
  "{\"entries\":[{\"id\":\"1\",\"hashParts\":[1,2,3,4]},{\"id\":\"5\",\"hashParts\":[1,2,3,5]},{\"id\":\"6\",\"hashParts\":[1,2,3,6]}]}",
  "{\"entries\":[{\"id\":\"1\",\"hashParts\":[1,2,3,4]},{\"id\":\"5\",\"hashParts\":[1,2,3,5]},{\"id\":\"6\",\"hashParts\":[1,2,3,6]}]}",
  "{\"entries\":[{\"id\":\"1\",\"hashParts\":[1,2,3,4]},{\"id\":\"5\",\"hashParts\":[1,2,3,5]},{\"id\":\"6\",\"hashParts\":[1,2,3,6]}]}"
]